File: solved.ac/gold/4/LCS 2.py
import sys
readline = sys.stdin.readline

A, B = readline().rstrip(), readline().rstrip()
Al,Bl = len(A),len(B)
#memorize
LCS = [[str() for _ in range(len(B)+1)] for _ in range(len(A)+1)]
for i in range(Al):
    for j in range(Bl):
        if A[i] == B[j]: LCS[i+1][j+1] = LCS[i][j] + A[i]
        else: LCS[i+1][j+1] = max(LCS[i][j+1], LCS[i+1][j], key=lambda x:len(x))
res1 = len(LCS[-1][-1])
res2 = LCS[-1][-1]
print(res1, res2, sep='\n')

File: solved.ac/gold/4/LCS 2.cpp
#include<iostream>
#include<vector>
using namespace std;

#define MAX(A,B) A.size() >= B.size() ? A : B

int main(void){
    string A,B; cin>>A>>B;
    int AL = A.size(), BL = B.size();
    vector<vector<string>> LCS(A.size()+1,vector<string>(B.size()+1));
    for (size_t i = 0; i < A.size(); i++)
    {
        for (size_t j = 0; j < B.size(); j++)
        {
            if(A[i] == B[i]) LCS[i+1][j+1] = LCS[i][j] + A[i];
            else LCS[i+1][j+1] = MAX(LCS[i][j+1], LCS[i+1][j]);
        }
    }
    cout << LCS[AL][BL] << '\n' << LCS[AL][BL].size();
}